### DISCUSSION TITLE
syntax highlighting on Alpine Linux didn't work

### DISCUSSION THREAD

**Discussion by DrSensor** [helix-editor/helix#9683] [Troubleshooting] [answered]
I've just updated my system and suddenly the syntax highlighting didn't work.

Previously it works but I need to install the grammar this way
```
apk add tree-sitter-{rust,toml}
```

**Comment by pascalkuthe** [comment]
The way that alpine package tree sitter grammars sepertely is not supported by us. We require grammars match the exact commit that we use in our `languages.toml` otherwise the queries may not work and other binary inconpatabilities are also possible.

We also don't maintain the alpine package (or any package manager package) so you need to report this to the maintainer of the alpine package and not upstream

**Comment by rdbo** [comment]
I was also having this issue. Installing the `tree-sitter-*` packages doesn't really solve the problem, because there are many, many missing languages.
It seems that there is a problem in the `helix-23.10-r2` package (I don't know exactly what the problem is, but from my surface level diagnosis, it seems that it's missing most of the `queries` in `/usr/share/helix/runtime`)
To fix this, downgrade helix to `23.10-r0`, present in the `v3.19` repos or lower than that:
```
echo 'http://dl-cdn.alpinelinux.org/alpine/v3.19/community' > /tmp/REPOS
sudo apk del helix
sudo apk add --no-cache --repositories-file=/tmp/REPOS helix
```
Now just fetch and build the grammars and you should be good to go:
```
hx -g fetch
hx -g build
```

**Reply by pascalkuthe** [reply]
Reporting this herr won't do any good. You need to report this to the alpine people (and also make it clear to them that packaging grammars separately is broken and not suplorted).

As upstream I can only advise to stay clear of the alone package and just install from source.

**Reply by rdbo** [reply]
Yeah, I reported this on the Alpine GitLab, just commented to let people know a temporary fix for this.

**Comment by gzmorell** [comment]
As a workaround, you can remove tree-sitter package and build tree-sitter libraries with the command "hx -g build" it should install the libraries in "$HOME/.config/helix/runtime/grammars"

**Reply by rdbo** [reply]
That won't work unfortunately, at least it didn't for me. It does fetch and build the libraries, but they don't get loaded (running `hx --health` shows that syntax highlighting for every language is not installed). The only way to fix it was downgrading the Alpine package, which seems to be broken, and only then build the tree-sitter libraries with `hx -g build`. I only managed to figure this out because I had previously installed helix on another machine and it would work there, then I attempted using that machine's version of helix on this one as well, and it worked.

**Comment by kodesoul** [comment]
Hello ! I've had the same problem. A fix that works is creating a queries folder at `.config/helix/runtime/queries`. Then manually copying the queries you need from `.config/helix/runtime/grammars/sources/<LANGUAGE>/queries` to `.config/helix/runtime/queries/<LANGUAGE>`. 

(P.S you do not need to have the tree-sitter-grammars alpine package installed).

**Comment by Carlo-Monte** [comment]
I can confirm that symlinking ~/.config/helix/runtime/grammars/sources/<LANG>/queries to ~/.config/helix/runtime/queries/<LANG> fixes the issue.

Can please someone from the Team explain, how exactly are those symlinks supposed to be created by a properly installed Helix? As far as I can see, "hx -g fetch" does a "git clone". It is not obvious why a git clone should specifically fail on Alpine, or why the build procedure from the repo should fail specifically on Alpine.

In order to be aber to blame Alpine, we first need to explain how exactly this is supposed to WORK CORRECTLY: who creates those symlinks normally??

**Reply by the-mikedavis** [reply]
You should not symlink `~/.config/helix/runtime/grammars/sources/<language>/queries` to `~/.config/helix/runtime/queries/<Language>`. The queries under `sources` belong to the grammar's repository (for example https://github.com/tree-sitter/tree-sitter-javascript/blob/master/queries/highlights.scm) and are meant for reference and testing. They might provide some highlighting but they almost never use [our scopes](https://docs.helix-editor.com/master/themes.html#syntax-highlighting) and usually have a different precedence (also see #9436) so the highlighting will be incomplete if it works at all. The queries under `runtime/queries/<language>` in the repository here are adapted for our capture names and precedence.

The alpine build should follow what other distributions do (for example Brew and Nix). As a reference the docs on building from source and setting up the runtime directory properly are here: https://docs.helix-editor.com/master/building-from-source.html#configuring-helixs-runtime-files

**Comment by Carlo-Monte** [comment]
Thank you, Alpine issue here: https://gitlab.alpinelinux.org/alpine/aports/-/issues/16336 .

I have one more question about the "Building from source" recipe. Does the Cargo file not provide an install target? Becuase the documentation strangely asks the user to copy stuff by hand: 
```
cp -r runtime $BUILD_DIR/usr/lib/helix/
cp target/opt/hx $BUILD_DIR/usr/bin/hx
```

**Reply by the-mikedavis** [reply]
That part of the installation instructions is meant to be a reference, package managers can install to whatever directory they want or use `cargo`'s `--root` option. That won't copy the runtime directory though, that needs to be `cp`'d anyway

**Reply by Carlo-Monte** [reply]
Thank you. Please close the issue.

**Comment by jgarte** [comment, marked as answer]
Hi,

What are these packages about?

https://pkgs.alpinelinux.org/packages?name=helix*&branch=edge&repo=&arch=x86_64&origin=&flagged=&maintainer=

 helix-tree-sitter-system 
 helix-tree-sitter-vendor

I couldn't find further documentation on them explaining why they exist. 

Will read the source code for the packages next to see what they do 🦆

**Reply by rdbo** [reply]
IIRC, system uses a system-wide tree-sitter and vendor uses the helix-specific one.
I use vendor without issues.

**Comment by Carlo-Monte** [comment]
Helix comes with its own customized Tree Sitter version and wont 
properly work with the original one.
"System" is the original package and "vendor" is the Tree Sitter 
customized for Helix.

In the Helix documentation, they require you to use the customized tree 
sitter.
In my opinion, until Helix is able to use the original external 
dependency, the customized Tree Sitter should be part of the Helix 
package.

Not only that people need to hunt for it and understand why Helix 
doesn't work with the default Tree Sitter,
but there is also the risk that someone else works in good faith on the 
packages and "cleans up".

I tried to convince the Helix package maintainer to include the 
customized tree sitter in the Helix package, but I had no success.
Please at least don't remove any of the two Tree Sitter packages: they 
are both needed.
